core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala

```diff
@@ -1919,7 +1919,12 @@ class TaskSetManagerSuite
     TestUtils.waitUntilExecutorsUp(sc, 2, 60000)
     val Seq(exec0, exec1) = backend.getExecutorIds()
 
-    val taskSet = FakeTask.createTaskSet(2)
+    // SPARK-30440: it's possible that tasks have already been scheduled after we call
+    // `sched.submitTasks(taskSet)` but before `manager.resourceOffer`. In this case,
```

Comment:
how can we schedule the task before receiving resource offers?

Reply:
Because this test is kind of somehow hacky. Instead of using mock  schedulerBackend and taskScheduler, we use the real `CoarseGrainedSchedulerBackend` and `TaskSchedulerImpl` from a `SparkContext` due to the reason that we want use `reset()` within `CoarseGrainedSchedulerBackend`. As a result, when we call `sched.submitTasks(taskSet)`, it will schedule tasks immediately.

I shall put this background in the comment, too.

Reply:
But how does changing number of tasks from 2 to 4 fix this.? as per your comment here, https://github.com/apache/spark/pull/27115#discussion_r363722012 this still remains flaky

Reply:
Because there's only two executors will be launched here. So, 2 tasks can be scheduled automatically, while other 2 tasks needs to be scheduled manually(a.k.a `resourceOffer`).

Reply:
There is no guarantee that the first two will not finish before calling `resourceOffer`, this fix just reduces the window of the failure but doesn't fix it. To resolve this scenario https://github.com/apache/spark/pull/27115 uses tasks which do not finish.

Reply:
You'r right as this point and we shall workaround it by increasing revive interval of `CoarseGrainedSchedulerBackend`.

As for your fix, I think it's kinda overkill but nothing else.

Reply:
I don't understand how is mine a overkill when actually what should matter is better approach for fix always. As i mentioned earlier, this PR duplicates development and review effort for https://github.com/apache/spark/pull/27115 then why still insist here.?